[InstCombine] Make foldCmpLoadFromIndexedGlobal more resiliant to non-array geps.

davemgreen opened the pull request:
My understanding is that gep [n x i8] and gep i8 can be treated equivalently - the array type conveys no extra information and could be removed. This goes through foldCmpLoadFromIndexedGlobal and tries to make it work for non-array gep types, so long as the index type still matches the array being loaded.

dtcxzyw commented:
IIRC https://github.com/llvm/llvm-project/pull/67093 is our final goal (i.e., cmp + load + gep i8/ptradd). But it looks stalled.

dtcxzyw reviewed (approved):
LG. It looks like we are still far from the type-independent solution. I think this patch may be a simple mitigation. As for regressions caused by this patch, we may avoid emitting multiple icmps when the load has other users.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -127,21 +133,27 @@ Instruction *InstCombinerImpl::foldCmpLoadFromIndexedGlobal(
+  // Require: GEP [n x i8] GV, 0, Idx {{, constant indices}}
+  //      Or: GEP i8 GV, Idx
+
+  SmallVector<unsigned, 4> LaterIndices;
```
Why move the declaration of LaterIndices?

nikic reviewed (approved):
I'm fine with this as well. This is not ideal, but still an improvement.

davemgreen committed:
[InstCombine] Make foldCmpLoadFromIndexedGlobal more resiliant to non-array geps.

My understanding is that gep [n x i8] and gep i8 can be treated equivalently -
the array type conveys no extra information and could be removed. This goes
through foldCmpLoadFromIndexedGlobal and tries to make it work for non-array
gep types, so long as the index type still matches the array being loaded.

davemgreen commented:
I should explain where this comes from, as there are a number of related issues going on at the same time. It was from a report I had of this code behaving differently between func1 and func2, just because of the different types of the GEP. One pulls an add out of the loop with rbit+ctz in it, the other doesn't. There are probably several ways to look at addressing that, but I figured the structs can be removed from the geps and that gets us closer to where we want to be.
https://godbolt.org/z/6dsK4zWhd

dtcxzyw commented:
> but I figured the structs can be removed from the geps and that gets us closer to where we want to be.

Yeah, we plan to strip leading zero indices. Then we can see if it breaks some existing transforms that only work with array source element types.

madhur13490 commented:
Hi @davemgreen 
Some tests are failing in one of our internal test suites. I have reduced the test to the IR: https://godbolt.org/z/a4W18rdsn

The phi node is getting replaced with a branch that eventually leads to an infinite loop. The test run faults with "Invalid Instruction" on Neoverse-V2 platform.

Can you please have a look and let me know if this can be fixed?
